cmd/tailscale/cli/up.go

```diff
@@ -635,8 +640,12 @@ func checkForAccidentalSettingReverts(flagSet map[string]bool, curPrefs *ipn.Pre
 				missing = append(missing, fmtFlagValueArg("exit-node", fmtSettingVal(exi)))
 			}
 		case "advertise-routes":
+			hadExitNode := hasExitNodeRoutes(exi.([]netaddr.IPPrefix))
 			routes := withoutExitNodes(exi.([]netaddr.IPPrefix))
 			missing = append(missing, fmtFlagValueArg("advertise-routes", fmtSettingVal(routes)))
+			if hadExitNode && !flagSet["advertise-exit-node"] {
```

Comment:
this hunk is the only function change (and the fix) in the PR.

The rest is just tests.